neo/SmartContract/Native/Tokens/GasToken.cs

```diff
@@ -0,0 +1,173 @@
+﻿using Neo.Ledger;
+using Neo.VM;
+using Neo.VM.Types;
+using System;
+using System.Linq;
+using System.Numerics;
+using System.Text;
+using VMArray = Neo.VM.Types.Array;
+
+namespace Neo.SmartContract.Native.Tokens
+{
+    public class GasToken : NativeContractBase
+    {
+        public const string ServiceName = "Neo.Native.Tokens.GAS";
+        public static readonly byte[] Script = CreateNativeScript(ServiceName);
+        public static readonly UInt160 ScriptHash = Script.ToScriptHash();
+        public static readonly string[] SupportedStandards = { "NEP-5", "NEP-10" };
+        public const string Name = "GAS";
+        public const string Symbol = "gas";
+        public const int Decimals = 8;
+        public static readonly BigInteger DecimalsFactor = BigInteger.Pow(10, Decimals);
+
+        internal static bool Main(ApplicationEngine engine)
+        {
+            if (!new UInt160(engine.CurrentContext.ScriptHash).Equals(ScriptHash))
+                return false;
+            string operation = engine.CurrentContext.EvaluationStack.Pop().GetString();
+            VMArray args = (VMArray)engine.CurrentContext.EvaluationStack.Pop();
+            StackItem result;
+            switch (operation)
+            {
+                case "supportedStandards":
+                    result = SupportedStandards.Select(p => (StackItem)p).ToList();
+                    break;
+                case "name":
+                    result = Name;
+                    break;
+                case "symbol":
+                    result = Symbol;
+                    break;
+                case "decimals":
+                    result = Decimals;
+                    break;
+                case "totalSupply":
+                    result = TotalSupply(engine);
+                    break;
+                case "balanceOf":
+                    result = BalanceOf(engine, args[0].GetByteArray());
+                    break;
+                case "transfer":
+                    result = Transfer(engine, args[0].GetByteArray(), args[1].GetByteArray(), args[2].GetBigInteger());
+                    break;
+                default:
+                    return false;
+            }
+            engine.CurrentContext.EvaluationStack.Push(result);
+            return true;
+        }
+
+        private static BigInteger TotalSupply(ApplicationEngine engine)
+        {
+            StorageItem storage = engine.Service.Snapshot.Storages.TryGet(new StorageKey
```

Comment:
Storage_Get method should be implemented with an option to receive String as parameter. So we don't keep using Encoding.ASCII stuff here, but simply ask for the base class to resolve this for us. This simplifies **a lot** this contract, and will be reused on many other scenarios.